File: tests/python/test_vtm.py
"""VTK XML MultiBlock (roadmap §1 tier B4, v11.6.0, part 3 of 3).

An index file plus one `.vtu` piece per cell block, combined on read via
`merge()` (no welding) into one mesh with one "cell" region per piece. Unlike
`.vts`/`.vtr`, there is no lattice restriction at all -- any mesh with one or
more cell blocks round-trips.
"""

import numpy as np
import pytest

import meshioplusplus
from meshioplusplus.vtm import _vtm


def _tri_quad_mesh():
    """Three blocks (triangle, quad, triangle) sharing points across
    blocks -- exercises point duplication across pieces and same-type block
    consolidation on the way back in, exactly like the C++ fixture."""
    points = np.array(
        [[0, 0, 0], [1, 0, 0], [2, 0, 0], [3, 1, 0], [2, 1, 0], [1, 1, 0], [0, 1, 0]],
        dtype=np.float64,
    )
    cells = [
        ("triangle", np.array([[0, 1, 5], [0, 5, 6]])),
        ("quad", np.array([[1, 2, 4, 5]])),
        ("triangle", np.array([[2, 3, 4]])),
    ]
    return meshioplusplus.Mesh(points, cells)


@pytest.mark.parametrize(
    "binary, compression", [(False, None), (True, None), (True, "zlib")]
)
def test_roundtrip(binary, compression, tmp_path):
    mesh = _tri_quad_mesh()
    path = tmp_path / "g.vtm"
    meshioplusplus.vtm.write(path, mesh, binary=binary, compression=compression)
    back = meshioplusplus.vtm.read(path)

    assert len(back.points) == 11  # 4 + 4 + 3, points duplicated across pieces
    assert [c.type for c in back.cells] == ["triangle", "quad"]
    assert len(back.cells[0].data) == 3  # the two triangle pieces consolidate
    assert len(back.cells[1].data) == 1

    assert [r.name for r in back.regions] == ["block_0", "block_1", "block_2"]
    assert all(r.kind == "cell" for r in back.regions)
    assert sum(len(r.entries) for r in back.regions) == 4


@pytest.mark.parametrize(
    "binary, compression", [(False, None), (True, None), (True, "zlib")]
)
def test_cross_compat(binary, compression, tmp_path):
    mesh = _tri_quad_mesh()
    cpp = tmp_path / "cpp.vtm"
    meshioplusplus.vtm.write(cpp, mesh, binary=binary, compression=compression)
    from_py = _vtm.read(cpp)
    assert len(from_py.points) == 11
    assert [r.name for r in from_py.regions] == ["block_0", "block_1", "block_2"]

    py = tmp_path / "py.vtm"
    _vtm.write(py, mesh, binary=binary, compression=compression)
    from_cpp = meshioplusplus.vtm.read(py)
    assert len(from_cpp.points) == 11
    assert [c.type for c in from_cpp.cells] == ["triangle", "quad"]


def test_generic_io(tmp_path):
    mesh = _tri_quad_mesh()
    path = tmp_path / "g.vtm"
    meshioplusplus.write(path, mesh)
    back = meshioplusplus.read(path)
    assert len(back.points) == 11
    assert meshioplusplus.sniff_format(path) == "vtm"


def test_pieces_are_independently_readable_vtu_files(tmp_path):
    mesh = _tri_quad_mesh()
    path = tmp_path / "g.vtm"
    meshioplusplus.vtm.write(path, mesh, binary=True)

    stem = path.stem
    expected_cells = [2, 1, 1]
    for i in range(3):
        piece = path.parent / stem / f"{stem}_{i}.vtu"
        assert piece.exists()
        p = meshioplusplus.read(piece)
        assert len(p.cells) == 1
        assert len(p.cells[0].data) == expected_cells[i]


def test_metadata_agrees_with_a_real_read(tmp_path):
    mesh = _tri_quad_mesh()
    path = tmp_path / "g.vtm"
    meshioplusplus.vtm.write(path, mesh, binary=True)

    meta = meshioplusplus.read_metadata(path)
    back = meshioplusplus.vtm.read(path)
    assert meta["num_points"] == len(back.points)
    assert len(meta["cell_blocks"]) == len(back.cells)
    for mb, cb in zip(meta["cell_blocks"], back.cells):
        assert mb["type"] == cb.type
        assert mb["num_cells"] == len(cb.data)


def test_declines_a_piece_that_is_neither_vtu_nor_vtp(tmp_path):
    path = tmp_path / "bad.vtm"
    path.write_text(
        '<?xml version="1.0"?>\n'
        '<VTKFile type="vtkMultiBlockDataSet" version="1.0">\n'
        "<vtkMultiBlockDataSet>\n"
        '<Block index="0">\n'
        '<DataSet index="0" name="a" file="piece.obj"/>\n'
        "</Block>\n"
        "</vtkMultiBlockDataSet>\n"
        "</VTKFile>\n"
    )
    with pytest.raises(meshioplusplus.ReadError):
        _vtm.read(path)


def test_empty_index_reads_as_an_empty_mesh(tmp_path):
    path = tmp_path / "empty.vtm"
    path.write_text(
        '<?xml version="1.0"?>\n'
        '<VTKFile type="vtkMultiBlockDataSet" version="1.0">\n'
        "<vtkMultiBlockDataSet>\n"
        '<Block index="0"/>\n'
        "</vtkMultiBlockDataSet>\n"
        "</VTKFile>\n"
    )
    for reader in (meshioplusplus.vtm.read, _vtm.read):
        back = reader(path)
        assert len(back.points) == 0
        assert len(back.cells) == 0

File: tests/cpp/test_vtm.cpp
//  ██████   ██████ ██████████  █████████  █████   █████ █████    ███████
// ░░██████ ██████ ░░███░░░░░█ ███░░░░░███░░███   ░░███ ░░███   ███░░░░░███      ███         ███
//  ░███░█████░███  ░███  █ ░ ░███    ░░░  ░███    ░███  ░███  ███     ░░███    ░███        ░███
//  ░███░░███ ░███  ░██████   ░░█████████  ░███████████  ░███ ░███      ░███ ███████████ ███████████
//  ░███ ░░░  ░███  ░███░░█    ░░░░░░░░███ ░███░░░░░███  ░███ ░███      ░███░░░░░███░░░ ░░░░░███░░░
//  ░███      ░███  ░███ ░   █ ███    ░███ ░███    ░███  ░███ ░░███     ███     ░███        ░███
//  █████     █████ ██████████░░█████████  █████   █████ █████ ░░░███████░      ░░░         ░░░
// ░░░░░     ░░░░░ ░░░░░░░░░░  ░░░░░░░░░  ░░░░░   ░░░░░ ░░░░░    ░░░░░░░
//
//
//  License:         MIT License
//                   meshio++ default license: LICENSE
//
//  Main authors:    Vicente Mataix Ferrandiz
//
//
// VTK XML MultiBlock (roadmap §1 tier B4, v11.6.0, part 3 of 3).

// System includes
#include <cstddef>
#include <filesystem>
#include <fstream>
#include <string>

// External includes
#include <gtest/gtest.h>

// Project includes
#include "mesh_fixtures.hpp"
#include "meshioplusplus/exceptions.hpp"
#include "meshioplusplus/formats/vtm.hpp"
#include "meshioplusplus/formats/vtu.hpp"
#include "meshioplusplus/region.hpp"

using meshioplusplus::Mesh;
using meshioplusplus::read_vtm;
using meshioplusplus::read_vtm_metadata;
using meshioplusplus::read_vtu;
using meshioplusplus::ReadError;
using meshioplusplus::RegionKind;
using meshioplusplus::write_vtm;

namespace {

// Remove the index file and its sibling pieces directory (`<stem>/`).
void vtm_cleanup(const std::string& rPath) {
    std::error_code ec;
    const std::filesystem::path p(rPath);
    std::filesystem::remove(p, ec);
    std::filesystem::remove_all(p.parent_path() / p.stem(), ec);
}

}  // namespace

// The roadmap's own stated probe: write two (here three) blocks to .vtm, read
// them back. `tri_quad_mesh()` has three blocks -- triangle, quad, triangle --
// sharing points across blocks, so this also exercises point duplication
// across pieces and same-type block consolidation on the way back in.
TEST(Vtm, RoundTripsThreeBlocksAsThreeNamedCellRegions) {
    const Mesh m = mt::tri_quad_mesh();
    const std::string path = mt::temp_path(".vtm");
    write_vtm(path, m, /*binary=*/false, /*zlib=*/false);
    const Mesh back = read_vtm(path);

    // Each piece is pruned to only the points its own block references:
    // {0,1,5,6} + {1,2,4,5} + {2,3,4} = 4 + 4 + 3, shared points duplicated.
    EXPECT_EQ(back.NumPoints(), 11u);
    // The two "triangle" pieces (index 0 and 2) consolidate into one block,
    // same as `merge()` would for any two same-typed inputs.
    ASSERT_EQ(back.NumCellBlocks(), 2u);
    EXPECT_EQ(back.Cells(0).Type(), "triangle");
    EXPECT_EQ(back.Cells(0).NumCells(), 3u);
    EXPECT_EQ(back.Cells(1).Type(), "quad");
    EXPECT_EQ(back.Cells(1).NumCells(), 1u);

    ASSERT_EQ(back.NumRegions(), 3u);
    std::size_t total_entries = 0;
    for (std::size_t i = 0; i < back.NumRegions(); ++i) {
        const meshioplusplus::Region& r = back.Region(i);
        EXPECT_EQ(r.mKind, RegionKind::Cell);
        EXPECT_EQ(r.mName, "block_" + std::to_string(i));
        total_entries += r.NumEntries();
    }
    EXPECT_EQ(total_entries, 4u);  // 2 + 1 + 1 original cells

    vtm_cleanup(path);
}

TEST(Vtm, PiecesAreIndependentlyReadableVtuFiles) {
    const Mesh m = mt::tri_quad_mesh();
    const std::string path = mt::temp_path(".vtm");
    write_vtm(path, m, /*binary=*/true, /*zlib=*/false);

    const std::filesystem::path idx(path);
    const std::filesystem::path dir = idx.parent_path() / idx.stem();
    const std::string stem = idx.stem().string();
    const std::size_t expected_cells[3] = {2, 1, 1};
    for (int i = 0; i < 3; ++i) {
        const std::string piece = (dir / (stem + "_" + std::to_string(i) + ".vtu")).string();
        ASSERT_TRUE(std::filesystem::exists(piece)) << piece;
        const Mesh p = read_vtu(piece);
        ASSERT_EQ(p.NumCellBlocks(), 1u);
        EXPECT_EQ(p.Cells(0).NumCells(), expected_cells[i]);
    }

    vtm_cleanup(path);
}

TEST(Vtm, MetadataAgreesWithARealRead) {
    const Mesh m = mt::tri_quad_mesh();
    const std::string path = mt::temp_path(".vtm");
    write_vtm(path, m, true, false);

    const meshioplusplus::MeshMetadata meta = read_vtm_metadata(path);
    const Mesh back = read_vtm(path);
    EXPECT_EQ(meta.mNumPoints, back.NumPoints());
    ASSERT_EQ(meta.mCellBlocks.size(), back.NumCellBlocks());
    for (std::size_t i = 0; i < meta.mCellBlocks.size(); ++i) {
        EXPECT_EQ(meta.mCellBlocks[i].mType, back.Cells(i).Type());
        EXPECT_EQ(meta.mCellBlocks[i].mNumCells, back.Cells(i).NumCells());
    }

    vtm_cleanup(path);
}

TEST(Vtm, EmptyIndexReadsAsAnEmptyMesh) {
    const std::string path = mt::temp_path(".vtm");
    std::ofstream os(path, std::ios::binary);
    os << "<?xml version=\"1.0\"?>\n"
          "<VTKFile type=\"vtkMultiBlockDataSet\" version=\"1.0\">\n"
          "<vtkMultiBlockDataSet>\n"
          "<Block index=\"0\"/>\n"
          "</vtkMultiBlockDataSet>\n"
          "</VTKFile>\n";
    os.close();

    const Mesh back = read_vtm(path);
    EXPECT_EQ(back.NumPoints(), 0u);
    EXPECT_EQ(back.NumCellBlocks(), 0u);

    const meshioplusplus::MeshMetadata meta = read_vtm_metadata(path);
    EXPECT_EQ(meta.mNumPoints, 0u);
    EXPECT_TRUE(meta.mCellBlocks.empty());

    std::filesystem::remove(path);
}

TEST(Vtm, DeclinesAPieceThatIsNeitherVtuNorVtp) {
    const std::string path = mt::temp_path(".vtm");
    std::ofstream os(path, std::ios::binary);
    os << "<?xml version=\"1.0\"?>\n"
          "<VTKFile type=\"vtkMultiBlockDataSet\" version=\"1.0\">\n"
          "<vtkMultiBlockDataSet>\n"
          "<Block index=\"0\">\n"
          "<DataSet index=\"0\" name=\"a\" file=\"piece.obj\"/>\n"
          "</Block>\n"
          "</vtkMultiBlockDataSet>\n"
          "</VTKFile>\n";
    os.close();

    EXPECT_THROW(read_vtm(path), ReadError);
    std::filesystem::remove(path);
}
